src/sinks/aws_cloudwatch_logs/config.rs

```diff
@@ -146,7 +148,8 @@ impl SinkConfig for CloudwatchLogsSinkConfig {
     }
 
     fn input(&self) -> Input {
-        Input::log()
+        let encoder_input_type = self.encoding.clone().config().input_type();
```

Comment:
Because `EncodingConfigAdapter::config` takes an owned `self`. The `clone` is on `encoding`, not on `input_type`. If `config` was to take `&self`, it would need to clone internally, so I decided to expose it on the outside.

> Not a big deal since this is only called once at start-up, but it just jumped out.

Agree, and for that reason I went with it rather than trying to do something more clever

Reply:
Ah, yeah, I see. It consumes itself to turn into a `config`? Would having an `as_config()` that returns a ref be feasible? Regardless, it's fine as-is.

Reply:
> Would having an `as_config()` that returns a ref be feasible?

Unfortunately not. There's two constraints here.

https://github.com/vectordotdev/vector/blob/abee634aa0f9b0eb136e65b3a62ccb1aa1ee6ac8/src/sinks/util/encoding/adapter.rs#L106-L112

We can not move out of `config`:

https://github.com/vectordotdev/vector/blob/abee634aa0f9b0eb136e65b3a62ccb1aa1ee6ac8/src/sinks/util/encoding/adapter.rs#L109

While at the same time we wouldn't be able to hand out a reference since we would reference a temporary result of `Migrator::migrate`:

https://github.com/vectordotdev/vector/blob/abee634aa0f9b0eb136e65b3a62ccb1aa1ee6ac8/src/sinks/util/encoding/adapter.rs#L110

Making `config` take `self` allows us not to clone again in places where the `encoding` is owned (this happens a couple of times in our code base). The trade off being that it requires to occasionally clone the `encoding` when we don't own it.

If you favor taking a reference and cloning the inner configs instead, we can do that. I'm not strongly in favor of one or the other, it just fell in place like that due to the given constraints

Reply:
Adressed in 2c8749009e169f85957abcc94064aa40e4a7849b